bottlerocket-os/bottlerocket
# testsys: Add declarative testing through `Test.toml`

**Issue number:**

Closes # N/A

**Description of changes:**

Enables configuring testing through `Test.toml`. There are now 5 levels of configuration.

```
# The test configuration is broken up into different levels depending on the variant.
# Each configuration's value will be selected in the following order:
# Arguments from the command line
# Environment variables from the command line
# '{variant}' configuration in `Test.toml`
# '{family}-{flavor}' configuration in `Test.toml`
# '{family}' configuration in `Test.toml`
# '{platform}-{flavor}' configuration in `Test.toml`
# '{platform}' configuration in `Test.toml`
# TestSys default values
```

**Testing done:**

Testing workflow over many different configurations.
* Create a sample config file
* Run `cargo make test`
* Use `kubectl describe test <TEST-NAME>` and `kubectl describe resource <RESOURCE-NAME>` to see the `configuration`
* Verify that the values in the `configuration` match the values in the config file.

I tested this with a variety of different configurations making sure that each field was selected properly for each possible field. I then tested that the configuration values were placed in the correct spots in the crds.

I also tested `Test.toml.example` with the redacted fields filled in to make sure that the documentation was correct. 


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.
